rig/utils/contexts.py

```diff
@@ -0,0 +1,117 @@
+"""Utilities for building contextual functions.
+
+The MachineController has a lot of functions which take the same arguments,
+many of which are contextual.  Avoiding respecifying these arguments every time
+will lead to cleaner and clearer code.  For example:
+
+    with controller(app_id=32):
+        with controller(x=1, y=1):
+            controller.do_something(...)
+
+        with controller(x=2, y=2):
+            controller.do_something(...)
+
+Is arguably clearer and less prone to silly mistakes than:
+
+    controller.do_something(x=1, y=1, app_id=32)
+    controller.do_something(x=2, y=2, app_id=32)
+
```

Comment:
"Though this form is still useful in other circumstances and thus should still be allowed."